src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -341,47 +353,82 @@ private Task InvokeStreamCore(string methodName, InvocationRequest irq, object[]
             }
         }
 
-        private async Task OnDataReceivedAsync(byte[] data)
+        private async Task StartReading()
         {
-            ResetTimeoutTimer();
-            if (_protocolReaderWriter.ReadMessages(data, _binder, out var messages))
+            try
             {
-                foreach (var message in messages)
+                while (true)
                 {
-                    InvocationRequest irq;
-                    switch (message)
+                    var result = await _connection.Input.ReadAsync();
+                    var buffer = result.Buffer;
+                    var consumed = buffer.End;
+                    var examined = buffer.End;
+
+                    try
                     {
-                        case InvocationMessage invocation:
-                            _logger.ReceivedInvocation(invocation.InvocationId, invocation.Target,
-                                invocation.ArgumentBindingException != null ? null : invocation.Arguments);
-                            await DispatchInvocationAsync(invocation, _connectionActive.Token);
-                            break;
-                        case CompletionMessage completion:
-                            if (!TryRemoveInvocation(completion.InvocationId, out irq))
-                            {
-                                _logger.DropCompletionMessage(completion.InvocationId);
-                                return;
-                            }
-                            DispatchInvocationCompletion(completion, irq);
-                            irq.Dispose();
-                            break;
-                        case StreamItemMessage streamItem:
-                            // Complete the invocation with an error, we don't support streaming (yet)
-                            if (!TryGetInvocation(streamItem.InvocationId, out irq))
+                        if (!buffer.IsEmpty)
+                        {
+                            ResetTimeoutTimer();
+
+                            if (_protocolReaderWriter.ReadMessages(buffer, _binder, out var messages, out consumed, out examined))
                             {
-                                _logger.DropStreamMessage(streamItem.InvocationId);
-                                return;
+                                foreach (var message in messages)
+                                {
+                                    InvocationRequest irq;
+                                    switch (message)
+                                    {
+                                        case InvocationMessage invocation:
```

Comment:
nit: Can we get some new lines in here

Reply:
This is literally copied from what was here before. Where do you want the newlines and does our other code look like that?